hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/upgrade/UpgradeDowngrade.java

```diff
@@ -275,4 +281,98 @@ protected Pair<Map<ConfigProperty, String>, List<ConfigProperty>> downgrade(Hood
+      return new SevenToEightUpgradeHandler();
+    } else if (fromVersion == HoodieTableVersion.EIGHT && toVersion == HoodieTableVersion.NINE) {
+      return new EightToNineUpgradeHandler();
+    } else {
+      throw new HoodieUpgradeDowngradeException(fromVersion.versionCode(), toVersion.versionCode(), true);
+    }
+  }
+
+  protected DowngradeHandler getDowngradeHandlerInstance(HoodieTableVersion fromVersion, HoodieTableVersion toVersion) {
+    if (fromVersion == HoodieTableVersion.ONE && toVersion == HoodieTableVersion.ZERO) {
+      return new OneToZeroDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.TWO && toVersion == HoodieTableVersion.ONE) {
+      return new TwoToOneDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.THREE && toVersion == HoodieTableVersion.TWO) {
+      return new ThreeToTwoDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.FOUR && toVersion == HoodieTableVersion.THREE) {
+      return new FourToThreeDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.FIVE && toVersion == HoodieTableVersion.FOUR) {
+      return new FiveToFourDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.SIX && toVersion == HoodieTableVersion.FIVE) {
+      return new SixToFiveDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.SEVEN && toVersion == HoodieTableVersion.SIX) {
+      return new SevenToSixDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.EIGHT && toVersion == HoodieTableVersion.SEVEN) {
+      return new EightToSevenDowngradeHandler();
+    } else if (fromVersion == HoodieTableVersion.NINE && toVersion == HoodieTableVersion.EIGHT) {
+      return new NineToEightDowngradeHandler();
+    } else {
+      throw new HoodieUpgradeDowngradeException(fromVersion.versionCode(), toVersion.versionCode(), false);
+    }
+  }
+
+  /**
+   * Checks if any handlers in the upgrade/downgrade path need rollback and compaction and performs it once before starting.
+   * This ensures rollback and compaction happens only when needed and only once at the very beginning of the process.
+   *
+   * @param fromVersion the current table version
+   * @param toVersion   the target table version
+   */
+  private void rollbackAndCompactIfNeeded(HoodieTableVersion fromVersion, HoodieTableVersion toVersion) {
+    // Check if any handlers in the upgrade/downgrade path need rollback and compaction
+    boolean needsRollbackAndCompact = false;
+    if (fromVersion.versionCode() < toVersion.versionCode()) {
+      // Check upgrade handlers
+      HoodieTableVersion checkVersion = fromVersion;
+      while (checkVersion.versionCode() < toVersion.versionCode()) {
+        HoodieTableVersion nextVersion = HoodieTableVersion.fromVersionCode(checkVersion.versionCode() + 1);
+        UpgradeHandler handler = getUpgradeHandlerInstance(checkVersion, nextVersion);
```

Comment:
do we really need to instantiate upgradeHandler instance here just to understand if we need to do rollback and compact. 
currently, I don't see much intelligence within a certain upgrade/downgrade handler based on which we do rollback pending and compact. 

if we are don't see much more logic in there, can we simplify this w/ just a set 

UPGRADE_HANDLER_REQUIRING_ROLLBACK_PENDING_COMMITS = {{7 -> 8}, {8 ->9}}
and just lookup in this set would suffice. 
CC @yihua

Reply:
I think i did this originally based on the feedback from here https://github.com/rahil-c/hudi/pull/2#discussion_r2234030450 around adding a `needsRollbackPendingCommitAndCompact` in each of the specific handlers that currently invoke it.

If you do not want this anymore i can switch over to just using the set approach idea like you mentioned, since i think that would eliminate alot of unnecessary code.

cc @nsivabalan @danny0405